File: worker/src/routes/play.ts
/// <reference types="@cloudflare/workers-types" />

import type { PlayEnv } from "../types";
import { sha256Hex } from "../utils/crypto";

interface StaticTarget {
  gameId: string;
  path: string;
  privateToken?: string;
}

interface PointerCacheEntry {
  prefix: string;
  expiresAt: number;
}

const sandboxPointerCache = new Map<string, PointerCacheEntry>();
const privateReleaseCache = new Map<string, PointerCacheEntry>();
const POINTER_CACHE_TTL_MS = 25_000; // 25s in-memory cache to eliminate D1 per-asset load bottle-neck

const MIME_TYPES: Record<string, string> = {
  wasm: "application/wasm",
  data: "application/octet-stream",
  unityweb: "application/octet-stream",
  pck: "application/octet-stream",
  js: "application/javascript; charset=utf-8",
  mjs: "application/javascript; charset=utf-8",
  json: "application/json",
  atlas: "application/json",
  css: "text/css; charset=utf-8",
  html: "text/html; charset=utf-8",
  htm: "text/html; charset=utf-8",
  txt: "text/plain; charset=utf-8",
  svg: "image/svg+xml",
  png: "image/png",
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  gif: "image/gif",
  webp: "image/webp",
  ico: "image/x-icon",
  mp3: "audio/mpeg",
  ogg: "audio/ogg",
  wav: "audio/wav",
  m4a: "audio/mp4",
  mp4: "video/mp4",
  webm: "video/webm",
  woff: "font/woff",
  woff2: "font/woff2",
  ttf: "font/ttf",
  otf: "font/otf",
  cwasm: "application/wasm",
  symbols: "application/octet-stream",
};

export async function handlePlayRequest(
  request: Request,
  env: PlayEnv,
  ctx?: ExecutionContext,
): Promise<Response | null> {
  const url = new URL(request.url);

  // Health check endpoint
  if (url.pathname === "/health" || url.pathname === "/api/health") {
    return new Response(JSON.stringify({ status: "ok", service: "gamecreator-play" }), {
      status: 200,
      headers: { "Content-Type": "application/json" },
    });
  }

  if (!env.ARTIFACTS || !["GET", "HEAD"].includes(request.method)) return null;

  // Cloudflare Edge Cache lookup for GET requests
  if (request.method === "GET") {
    const cache = caches.default;
    const cached = await cache.match(request);
    if (cached) return cached;
  }

  const target = resolveStaticTarget(url, env);
  if (!target) return null;

  // Look up release prefix (in-memory cached or D1)
  const prefix = target.privateToken
    ? await resolvePrivateReleasePrefix(target.gameId, target.privateToken, env)
    : await resolveSandboxReleasePrefix(target.gameId, env);

  if (!prefix) return new Response("Game release not found", { status: 404 });

  let objectKey = `${prefix}/${target.path}`;
  let object = request.method === "HEAD"
    ? await env.ARTIFACTS.head(objectKey)
    : await env.ARTIFACTS.get(objectKey);

  let isPrecompressedBr = false;
  // If not found, check if a .br precompressed version was uploaded (e.g. game.wasm.br)
  const acceptEncoding = request.headers.get("Accept-Encoding") || "";
  if (!object && acceptEncoding.includes("br") && !target.path.endsWith(".br")) {
    const brKey = `${objectKey}.br`;
    const brObject = request.method === "HEAD"
      ? await env.ARTIFACTS.head(brKey)
      : await env.ARTIFACTS.get(brKey);
    if (brObject) {
      object = brObject;
      objectKey = brKey;
      isPrecompressedBr = true;
    }
  }

  // SPA fallback: if not found, and target.path is a clean route without dot (e.g. /levels/1), try index.html
  if (!object && target.path !== "index.html" && !target.path.includes(".")) {
    const fallbackKey = `${prefix}/index.html`;
    object = request.method === "HEAD"
      ? await env.ARTIFACTS.head(fallbackKey)
      : await env.ARTIFACTS.get(fallbackKey);
    if (object) {
      target.path = "index.html";
      objectKey = fallbackKey;
    }
  }

  if (!object) return new Response("File not found", { status: 404 });

  const headers = new Headers();
  object.writeHttpMetadata(headers);
  headers.set("ETag", object.httpEtag);

  // Determine MIME type accurately for WASM, WebGL, Unity, Godot, etc.
  const mimeType = getMimeType(target.path);
  if (mimeType) {
    headers.set("Content-Type", mimeType);
  }

  if (isPrecompressedBr || target.path.endsWith(".br")) {
    headers.set("Content-Encoding", "br");
  } else if (target.path.endsWith(".gz")) {
    headers.set("Content-Encoding", "gzip");
  }

  const isImmutableAsset = target.path !== "index.html";
  headers.set(
    "Cache-Control",
    isImmutableAsset
      ? "public, max-age=31536000, immutable"
      : "no-cache, no-store, must-revalidate",
  );

  applyGameSecurityHeaders(headers, env);

  const response = new Response(
    request.method === "HEAD" ? null : object.body,
    { status: 200, headers },
  );

  // Store in Cloudflare Edge Cache for immutable assets
  if (request.method === "GET" && isImmutableAsset && ctx) {
    ctx.waitUntil(caches.default.put(request, response.clone()));
  }

  return response;
}

function resolveStaticTarget(url: URL, env: PlayEnv): StaticTarget | null {
  const parts = url.pathname.split("/").filter(Boolean);
  if (parts.length === 0) return null;

  const queryToken = url.searchParams.get("token") || undefined;

  // 1. Optional subdomain mapping (if custom domain is mapped in the future)
  const hostname = url.hostname.toLowerCase();
  const playDomain = (env.PLAY_BASE_DOMAIN || "").toLowerCase();
  if (playDomain && hostname.endsWith(`.${playDomain}`)) {
    const gameId = hostname.slice(0, -(playDomain.length + 1));
    if (isSafeGameHostLabel(gameId)) {
      if (parts[0] === "private" && parts[1]) {
        const path = normalizeFilePath(parts.slice(2).join("/") || "index.html");
        return path ? { gameId, path, privateToken: decodeURIComponent(parts[1]) } : null;
      }
      const path = normalizeFilePath(parts.join("/") || "index.html");
      return path ? { gameId, path, privateToken: queryToken } : null;
    }
  }

  // 2. Path-based routing (used on workers.dev or direct paths)
  let gameId: string;
  let subParts: string[];

  if (parts[0] === "sandbox") {
    if (!parts[1]) return null;
    gameId = decodeURIComponent(parts[1]);
    subParts = parts.slice(2);
  } else {
    gameId = decodeURIComponent(parts[0]);
    subParts = parts.slice(1);
  }

  if (!isSafeGameHostLabel(gameId)) return null;

  if (subParts[0] === "private" && subParts[1]) {
    const privateToken = decodeURIComponent(subParts[1]);
    const path = normalizeFilePath(subParts.slice(2).join("/") || "index.html");
    return path ? { gameId, path, privateToken } : null;
  }

  const path = normalizeFilePath(subParts.join("/") || "index.html");
  return path ? { gameId, path, privateToken: queryToken } : null;
}

function isSafeGameHostLabel(value: string): boolean {
  return /^[a-z0-9][a-z0-9_-]{0,62}$/i.test(value);
}

function normalizeFilePath(path: string): string | null {
  if (
    !path ||
    path.includes("\\") ||
    path.startsWith("/") ||
    path.split("/").some((part) => !part || part === "." || part === "..")
  ) {
    return null;
  }
  return path.length <= 512 ? path : null;
}

function getMimeType(path: string): string | null {
  const cleanPath = path.endsWith(".br") || path.endsWith(".gz")
    ? path.slice(0, path.lastIndexOf("."))
    : path;
  const ext = cleanPath.split(".").pop()?.toLowerCase();
  return (ext && MIME_TYPES[ext]) || null;
}

async function resolveSandboxReleasePrefix(gameId: string, env: PlayEnv): Promise<string | null> {
  const now = Date.now();
  const cached = sandboxPointerCache.get(gameId);
  if (cached && cached.expiresAt > now) {
    return cached.prefix;
  }

  const row = await env.DB.prepare(
    `SELECT artifact_prefix FROM game_release_pointers WHERE game_id = ?`,
  ).bind(gameId).first<{ artifact_prefix: string }>();

  if (!row?.artifact_prefix) return null;

  sandboxPointerCache.set(gameId, {
    prefix: row.artifact_prefix,
    expiresAt: now + POINTER_CACHE_TTL_MS,
  });

  return row.artifact_prefix;
}

async function resolvePrivateReleasePrefix(
  gameId: string,
  token: string,
  env: PlayEnv,
): Promise<string | null> {
  const tokenHash = await sha256Hex(token);
  const cacheKey = `${gameId}:${tokenHash}`;
  const now = Date.now();
  const cached = privateReleaseCache.get(cacheKey);
  if (cached && cached.expiresAt > now) {
    return cached.prefix;
  }

  const row = await env.DB.prepare(
    `SELECT d.artifact_prefix
     FROM private_releases r
     JOIN deployment_records d ON d.id = r.deployment_id
     WHERE r.game_id = ? AND r.token_hash = ? AND r.revoked_at IS NULL
       AND (r.expires_at IS NULL OR r.expires_at > ?)
       AND d.status = 'published'`,
  ).bind(gameId, tokenHash, now).first<{ artifact_prefix: string }>();

  if (!row?.artifact_prefix) return null;

  privateReleaseCache.set(cacheKey, {
    prefix: row.artifact_prefix,
    expiresAt: now + POINTER_CACHE_TTL_MS,
  });

  return row.artifact_prefix;
}

function applyGameSecurityHeaders(headers: Headers, env: PlayEnv): void {
  const connectSources = (env.GAME_CSP_CONNECT_SRC || "'self'")
    .split(",")
    .map((value) => value.trim())
    .filter(Boolean)
    .join(" ");

  // Allow embedding in randseed.org, localhost for development, and devcreator.randseed.org
  headers.set(
    "Content-Security-Policy",
    `default-src 'self' blob: data:; base-uri 'none'; object-src 'none'; frame-ancestors 'self' https://randseed.org https://*.randseed.org https://devcreator.randseed.org http://localhost:* http://127.0.0.1:*; form-action 'none'; script-src 'self' 'unsafe-eval' 'unsafe-inline' 'wasm-unsafe-eval' blob:; style-src 'self' 'unsafe-inline'; img-src 'self' data: blob: https:; media-src 'self' data: blob: https:; font-src 'self' data: https:; connect-src 'self' ${connectSources} https: wss:; worker-src 'self' blob:; manifest-src 'self'`,
  );
  headers.set("X-Content-Type-Options", "nosniff");
  headers.set("Referrer-Policy", "no-referrer-when-downgrade");
  headers.set("Permissions-Policy", "camera=(), microphone=(), geolocation=(), autoplay=(self), fullscreen=(self), gamepad=(self)");
  headers.set("Cross-Origin-Opener-Policy", "same-origin");
  headers.set("Cross-Origin-Embedder-Policy", "credentialless");
  headers.set("Cross-Origin-Resource-Policy", "cross-origin");
  headers.set("Access-Control-Allow-Origin", "*");
  headers.set("Access-Control-Allow-Methods", "GET, HEAD, OPTIONS");
}

File: rust-worker/src/play.rs
use std::{cell::RefCell, collections::HashMap};

use serde_json::json;
use sha2::{Digest, Sha256};
use worker::{js_sys, Bucket, Cache, Env, Method, Request, Response, Result};

use crate::db;

const POINTER_CACHE_TTL_MS: i64 = 25_000;

#[derive(Clone)]
struct PointerCacheEntry {
    prefix: String,
    expires_at: i64,
}

thread_local! {
    static POINTER_CACHE: RefCell<HashMap<String, PointerCacheEntry>> = RefCell::new(HashMap::new());
}

pub async fn route(request: &Request, env: &Env) -> Result<Option<Response>> {
    if !matches!(request.method(), Method::Get | Method::Head) { return Ok(None); }
    let path = request.path();
    let parts: Vec<&str> = path.split('/').filter(|value| !value.is_empty()).collect();
    if parts.is_empty() || parts[0].starts_with("api") || parts[0] == "dashboard" { return Ok(None); }
    let mut game_id = parts[0].to_string();
    let mut index = 1;
    let mut private_token = None;
    if parts[0] == "sandbox" && parts.len() > 1 { game_id = parts[1].to_string(); index = 2; }
    else if parts[0] == "private" && parts.len() > 2 { game_id = parts[1].to_string(); private_token = Some(parts[2].to_string()); index = 3; }
    let url = request.url()?;
    if private_token.is_none() { private_token = url.query_pairs().find(|(key, _)| key == "token").map(|(_, value)| value.to_string()); }
    let mut target_path = normalize_path(parts.get(index..).unwrap_or_default().join("/").as_str()).unwrap_or_else(|| "index.html".to_string());
    let is_sandbox = parts[0] == "sandbox";
    let is_private = parts[0] == "private" || private_token.is_some();
    if !is_sandbox && !is_private { game_id = resolve_public_game_id(&game_id, env).await?; }
    let database = db::database(env)?;
    let prefix = if let Some(token) = private_token { resolve_private_prefix(&database, &game_id, &token).await? } else if is_sandbox { resolve_sandbox_prefix(&database, &game_id).await? } else { resolve_public_prefix(&database, &game_id).await? };
    let Some(prefix) = prefix else { return Ok(Some(Response::error("Game release not found", 404)?)); };
    let cache_request = if request.method() == Method::Get && !is_private && target_path != "index.html" {
        let separator = if url.query_pairs().next().is_some() { '&' } else { '?' };
        Some(Request::new(&format!("{}{}__release={}", request.url()?, separator, urlencoding::encode(&prefix)), Method::Get)?)
    } else { None };
    if let Some(cache_request) = cache_request.as_ref() {
        if let Some(response) = Cache::open("default".to_string()).await.get(cache_request, true).await? { return Ok(Some(response)); }
    }
    let bucket: Bucket = env.bucket("ARTIFACTS")?;
    let mut key = format!("{prefix}/{target_path}");
    let mut object = if request.method() == Method::Head { bucket.head(&key).await? } else { bucket.get(&key).execute().await? };
    if object.is_none() && request.headers().get("Accept-Encoding").ok().flatten().unwrap_or_default().contains("br") && !target_path.ends_with(".br") { key.push_str(".br"); object = if request.method() == Method::Head { bucket.head(&key).await? } else { bucket.get(&key).execute().await? }; }
    if object.is_none() && target_path != "index.html" && !target_path.contains('.') {
        key = format!("{prefix}/index.html");
        object = if request.method() == Method::Head { bucket.head(&key).await? } else { bucket.get(&key).execute().await? };
        if object.is_some() { target_path = "index.html".to_string(); }
    }
    if object.is_none() { return Ok(Some(Response::error("File not found", 404)?)); }
    let Some(object) = object else { return Ok(Some(Response::error("File not found", 404)?)); };
    let connect_sources = env.var("GAME_CSP_CONNECT_SRC").map(|value| value.to_string()).unwrap_or_else(|_| "'self'".to_string()).split(',').map(str::trim).filter(|value| !value.is_empty()).collect::<Vec<_>>().join(" ");
    let csp = format!("default-src 'self' blob: data:; base-uri 'none'; object-src 'none'; frame-ancestors 'self' https://randseed.org https://*.randseed.org https://devcreator.randseed.org http://localhost:* http://127.0.0.1:*; form-action 'none'; script-src 'self' 'unsafe-eval' 'unsafe-inline' 'wasm-unsafe-eval' blob:; style-src 'self' 'unsafe-inline'; img-src 'self' data: blob: https:; media-src 'self' data: blob: https:; font-src 'self' data: https:; connect-src 'self' {connect_sources} https: wss:; worker-src 'self' blob:; manifest-src 'self'");
    let mut builder = Response::builder().with_header("Cache-Control", if target_path == "index.html" { "no-cache, no-store, must-revalidate" } else { "public, max-age=31536000, immutable" })?.with_header("Content-Type", mime(&target_path))?.with_header("Content-Security-Policy", &csp)?.with_header("X-Content-Type-Options", "nosniff")?.with_header("Referrer-Policy", "no-referrer-when-downgrade")?.with_header("Permissions-Policy", "camera=(), microphone=(), geolocation=(), autoplay=(self), fullscreen=(self), gamepad=(self)")?.with_header("Cross-Origin-Opener-Policy", "same-origin")?.with_header("Cross-Origin-Embedder-Policy", "credentialless")?.with_header("Cross-Origin-Resource-Policy", "cross-origin")?;
    if key.ends_with(".br") { builder = builder.with_header("Content-Encoding", "br")?; }
    if request.method() == Method::Head { return Ok(Some(builder.empty())); }
    let body = object.body().ok_or_else(|| worker::Error::from("release body unavailable"))?.response_body()?;
    let response = builder.body(body);
    if let Some(cache_request) = cache_request.as_ref() {
        let cache = Cache::open("default".to_string()).await;
        cache.put(cache_request, response).await?;
        return Ok(cache.get(cache_request, true).await?);
    }
    Ok(Some(response))
}

async fn resolve_public_game_id(identifier: &str, env: &Env) -> Result<String> {
    let cache_key = format!("public:{identifier}").to_ascii_lowercase();
    if let Some(value) = cached_prefix(&cache_key) { return Ok(value); }
    let row = db::first(&db::database(env)?, "SELECT id FROM games WHERE lower(trim(short_name)) = lower(trim(?)) AND status = 'PUBLIC_ACTIVE' LIMIT 1", &[json!(identifier)]).await?;
    let resolved = row.and_then(|value| db::string(&value, "id")).unwrap_or_else(|| identifier.to_string());
    cache_prefix(cache_key, resolved.clone());
    Ok(resolved)
}

async fn resolve_public_prefix(database: &worker::d1::D1Database, game_id: &str) -> Result<Option<String>> {
    let cache_key = format!("pointer:{game_id}");
    if let Some(value) = cached_prefix(&cache_key) { return Ok(Some(value)); }
    let prefix = db::first(database, "SELECT p.artifact_prefix FROM game_release_pointers p JOIN games g ON g.id = p.game_id WHERE p.game_id = ? AND g.status = 'PUBLIC_ACTIVE'", &[json!(game_id)]).await?.and_then(|row| db::string(&row, "artifact_prefix"));
    if let Some(value) = prefix.clone() { cache_prefix(cache_key, value); }
    Ok(prefix)
}

async fn resolve_sandbox_prefix(database: &worker::d1::D1Database, game_id: &str) -> Result<Option<String>> {
    let cache_key = format!("sandbox:{game_id}");
    if let Some(value) = cached_prefix(&cache_key) { return Ok(Some(value)); }
    let prefix = db::first(database, "SELECT artifact_prefix FROM game_sandbox_pointers WHERE game_id = ?", &[json!(game_id)]).await?.and_then(|row| db::string(&row, "artifact_prefix"));
    if let Some(value) = prefix.clone() { cache_prefix(cache_key, value); }
    Ok(prefix)
}

async fn resolve_private_prefix(database: &worker::d1::D1Database, game_id: &str, token: &str) -> Result<Option<String>> {
    let hash = hex::encode(Sha256::digest(token.as_bytes()));
    let prefix = db::first(database, "SELECT d.artifact_prefix FROM private_releases r JOIN deployment_records d ON d.id = r.deployment_id WHERE r.game_id = ? AND r.token_hash = ? AND r.revoked_at IS NULL AND (r.expires_at IS NULL OR r.expires_at > ?) AND d.status IN ('ready', 'published')", &[json!(game_id), json!(hash), json!(js_sys::Date::now() as i64)]).await?.and_then(|row| db::string(&row, "artifact_prefix"));
    Ok(prefix)
}

fn cached_prefix(key: &str) -> Option<String> {
    let now = js_sys::Date::now() as i64;
    POINTER_CACHE.with(|cache| {
        let mut cache = cache.borrow_mut();
        let entry = cache.get(key).cloned();
        if entry.as_ref().map(|value| value.expires_at > now).unwrap_or(false) { return entry.map(|value| value.prefix); }
        cache.remove(key);
        None
    })
}

fn cache_prefix(key: String, prefix: String) {
    POINTER_CACHE.with(|cache| { cache.borrow_mut().insert(key, PointerCacheEntry { prefix, expires_at: js_sys::Date::now() as i64 + POINTER_CACHE_TTL_MS }); });
}

fn normalize_path(path: &str) -> Option<String> { if path.is_empty() { return Some("index.html".to_string()); } if path.len() > 512 || path.starts_with('/') || path.contains('\\') || path.split('/').any(|part| part.is_empty() || part == "." || part == "..") { None } else { Some(path.to_string()) } }
fn mime(path: &str) -> &'static str { match path.rsplit('.').next().unwrap_or_default().to_ascii_lowercase().as_str() { "html" => "text/html; charset=utf-8", "js" | "mjs" => "application/javascript; charset=utf-8", "css" => "text/css; charset=utf-8", "json" => "application/json", "wasm" => "application/wasm", "png" => "image/png", "jpg" | "jpeg" => "image/jpeg", "svg" => "image/svg+xml", "mp4" => "video/mp4", _ => "application/octet-stream" } }